cmd/gcp/omniwitness/persistence.go

```diff
@@ -90,24 +99,40 @@ func (p *spannerPersistence) createTablesIfNotExist(ctx context.Context) error {
 	return nil
 }
 
-func (p *spannerPersistence) AddLogs(ctx context.Context, lc omniwitness.ConfigYAML) error {
-	_, err := p.spanner.ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
-		m := []*spanner.Mutation{}
-		for _, l := range lc.Logs {
-			m = append(m, spanner.InsertOrUpdate(
-				"logs",
-				[]string{"logID", "origin", "vkey", "url", "feeder"},
-				[]any{logfmt.ID(l.Origin), l.Origin, l.PublicKey, l.URL, l.Feeder.String()},
-			))
+func (p *spannerPersistence) AddLogs(ctx context.Context, lc []config.Log) error {
+	m := []*spanner.MutationGroup{}
+	for _, l := range lc {
+		m = append(m, &spanner.MutationGroup{Mutations: []*spanner.Mutation{spanner.Insert(
+			"logs",
+			[]string{"logID", "origin", "vkey", "contact"},
+			[]any{logfmt.ID(l.Origin), l.Origin, l.VKey, l.Contact},
+		)}})
+	}
+	return p.batchWrite(ctx, p.spanner, m)
+}
+
+func batchWrite(ctx context.Context, s *spanner.Client, m []*spanner.MutationGroup) error {
+	errs := []error{}
+	err := s.BatchWrite(ctx, m).Do(func(r *spannerpb.BatchWriteResponse) error {
+		switch c := codes.Code(r.Status.Code); c {
+		case codes.OK, codes.AlreadyExists:
```

Comment:
Based on https://cloud.google.com/spanner/docs/error-codes, I think that AlreadyExists will be raised if there is already an entry with the same key, but other fields might have changed.